core/src/main/scala/kafka/server/DelayedRemoteListOffsets.scala

```diff
@@ -0,0 +1,174 @@
+    status.completed = status.futureHolderOpt.isEmpty
+    if (status.futureHolderOpt.isDefined) {
+      status.responseOpt = Some(buildErrorResponse(Errors.REQUEST_TIMED_OUT, topicPartition.partition()))
+    }
+    trace(s"Initial partition status for $topicPartition is $status")
+  }
+
+  /**
+   * Call-back to execute when a delayed operation gets expired and hence forced to complete.
+   */
+  override def onExpiration(): Unit = {
+    metadata.statusByPartition.forKeyValue { (topicPartition, status) =>
+      if (!status.completed) {
+        debug(s"Expiring list offset request for partition $topicPartition with status $status")
+        status.futureHolderOpt.foreach(futureHolder => futureHolder.jobFuture.cancel(true))
+        DelayedRemoteListOffsetsMetrics.recordExpiration(topicPartition)
+      }
+    }
+  }
+
+  /**
+   * Process for completing an operation; This function needs to be defined
+   * in subclasses and will be called exactly once in forceComplete()
+   */
+  override def onComplete(): Unit = {
+    val responseTopics = metadata.statusByPartition.groupBy(e => e._1.topic()).map {
+      case (topic, status) =>
+        new ListOffsetsTopicResponse().setName(topic).setPartitions(status.values.flatMap(s => s.responseOpt).toList.asJava)
+    }.toList
+    responseCallback(responseTopics)
+  }
+
+  /**
+   * Try to complete the delayed operation by first checking if the operation
+   * can be completed by now. If yes execute the completion logic by calling
+   * forceComplete() and return true iff forceComplete returns true; otherwise return false
+   *
+   * This function needs to be defined in subclasses
+   */
+  override def tryComplete(): Boolean = {
+    var completable = true
+    metadata.statusByPartition.forKeyValue { (partition, status) =>
+      if (!status.completed) {
+        status.futureHolderOpt.foreach { futureHolder =>
+          if (futureHolder.taskFuture.isDone) {
+            val response = futureHolder.taskFuture.get() match {
+              case Left(e) =>
+                buildErrorResponse(Errors.forException(e), partition.partition())
+
+              case Right(None) =>
+                val error = status.maybeOffsetsError
+                  .map(e => if (version >= 5) Errors.forException(e) else Errors.LEADER_NOT_AVAILABLE)
+                  .getOrElse(Errors.NONE)
+                buildErrorResponse(error, partition.partition())
+
+              case Right(Some(found)) =>
+                var partitionResponse = buildErrorResponse(Errors.NONE, partition.partition())
+                if (status.lastFetchableOffset.isDefined && found.offset >= status.lastFetchableOffset.get) {
+                  if (status.maybeOffsetsError.isDefined) {
+                    val error = if (version >= 5) Errors.forException(status.maybeOffsetsError.get) else Errors.LEADER_NOT_AVAILABLE
+                    partitionResponse.setErrorCode(error.code())
+                  }
```

Comment:
It's weird there's no `else` case when lastFetchableOffset <= found offset. From the original logic, we  should still return error, right?

Reply:
The original logic returns `empty` when  lastFetchableOffset <= found and maybeOffsetsError is undefined. If the result is empty, then we return (no-error, -1, -1L) from KafkaApis.scala:

1. https://sourcegraph.com/github.com/apache/kafka@049b7cde4c2f75a6af164d7308318e29878a2eca/-/blob/core/src/main/scala/kafka/cluster/Partition.scala?L1617
2. https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/core/src/main/scala/kafka/server/KafkaApis.scala?L1218